### ISSUE TITLE
Error with non-ASCII timezone names

### ISSUE THREAD

**Issue by jason-s** [dateutil/dateutil#809]
https://github.com/dateutil/dateutil/blob/master/dateutil/tz/_common.py#L22

```
def tzname_in_python2(namefunc):
    """Change unicode output into bytestrings in Python 2
    tzname() API changed in Python 3. It used to return bytes, but was changed
    to unicode strings
    """
    def adjust_encoding(*args, **kwargs):
        name = namefunc(*args, **kwargs)
        if name is not None and not PY3:
            name = name.encode()

        return name

    return adjust_encoding
```

The `encode` method will fail with non-ASCII timezone names.

**Comment by jason-s** [comment]
I have debugged this on a colleague's computer and the `name` string that causes this problem is '\xc3\xc0\xb9\xfa\xc9\xbd\xb5\xd8\xb1\xea\xd7\xbc\xca\xb1\xbc\xe4'

**Comment by jason-s** [comment]
OK, this is apparently the `gbk` codec: 

```
print '\xc3\xc0\xb9\xfa\xc9\xbd\xb5\xd8\xb1\xea\xd7\xbc\xca\xb1\xbc\xe4'.decode('gbk')
```

which prints as 美国山地标准时间 (Google Translate says it's American Mountain Standard Time)

**Comment by pganssle** [comment]
I'm fairly certain this is at least related to #416.

One issue here is that this is not actually a bug in `dateutil`, this is the result of a misconfigured locale or something. The *correct* solution here is to stop using Python 2, because Python 3 has *much* better defaults than Python 2 - which is why this is not an issue.

As far as I understand, the problem is that the way to fix this is to make sure your default string encoding (e.g. what gets used when you call `str.encode`) is compatible with the encoding of the time zone string. We could probably switch it to `utf-8` because that will be compatible with *most* things, but if someone is using time zone strings in an encoding that is *not* compatible with `utf-8`, they would have no way to specify this.

That said, there *may* be a way to get it to "just work" in more cases, by doing something like this:

```python
def tzname_in_python2(namefunc):
    """Change unicode output into bytestrings in Python 2
    tzname() API changed in Python 3. It used to return bytes, but was changed
    to unicode strings
    """
    def adjust_encoding(*args, **kwargs):
        name = namefunc(*args, **kwargs)
        if name is not None and not PY3:
            try:
                name = name.encode()
            except UnicodeDecodeError: 
                name = name.encode('utf-8', 'backslashreplace')
                msg = (u"Warning: failed to encode time zone name %s in the " +
                       u"default locale, falling back to utf-8 with " +
                       u"backslash replacement.") % name
                warnings.warn(msg, RuntimeWarning)

        return name

    return adjust_encoding
```

That said, because this is not a bug such as a "Python 2 is annoying" situation, I'm reluctant to spend much effort on it. We're going to be dropping Python 2 support relatively soon, and this already "just works" in Python 3.

**Comment by pganssle** [comment]
I would accept a PR for this, but keep in mind that the difficult part of any PR of this sort will be writing a test that can reproduce the locale issues with minimal overhead, preferably in a context manager. Since this locale stuff is always a total pain for me to test locally, any PR without proper tests would probably not be accepted.

**Comment by jbrockmendel** [comment]
> preferably in a context manager. 

@jason-s you might want to look at `pandas.util.testing.set_locale` for this

**Comment by dwoz** [comment]
This line is trying to `encode` a byte string to unicode. It should be a `decode` if the method should return unicode, no?

On windows, time.tzname is encoded with `mbcs`. See this upstream python issue:
https://bugs.python.org/issue16322

**Comment by pganssle** [comment]
> This line is trying to `encode` a byte string to unicode. It should be a `decode` if the method should return unicode, no?

This decorator is only applied in Python 2, and the `tzinfo` interface for Python 2 expects to return `bytes` (which is to say that `tzname` always returns `str`, however that's defined in the language). All of `dateutil`'s `tzname` methods are written to return unicode objects and this decorator converts them to the appropriate type for the language version.

**Comment by dwoz** [comment]
> 
> This decorator is only applied in Python 2, and the `tzinfo` interface for Python 2 expects to return `bytes` (which is to say that `tzname` always returns `str`, however that's defined in the language). All of `dateutil`'s `tzname` methods are written to return unicode objects and this decorator converts them to the appropriate type for the language version.

Yes, this is what I am saying. We want to return unicode but are doing an `encode` on bytes. `encode` calls are to encode unicode strings. Decode is used to return unicode from encoded bytes, which `tzinfo` has on python2
